kb/2024-11-28/kubevirt_cert_rotation.md

```diff
@@ -0,0 +1,58 @@
+---
+title: KubeVirt Certificates Rotation
+description: KubeVirt Certificates Rotation.
+slug: kubevirt_certificates_rotation
+authors:
+  - name: Cooper Tseng
+    title: Software Engineer
+    url: https://github.com/brandboat
+    image_url: https://github.com/brandboat.png
+tags: [harvester, kubevirt, certificates, cert, ca]
+hide_table_of_contents: false
+---
+
+- Harvester: [Issue 5798](https://github.com/harvester/harvester/issues/5798)
+
+In Harvester's embedded Rancher UI, you may see some warnings on the secrets page indicating that the KubeVirt certicates are about to expire. While there is no need to worry about it as KubeVirt handles automatic certificate rotation. 
+
+![kubevirt-certs-expired](./imgs/kubevirt_certs_expired.png)
+
+This page explains how KubeVirt manages self-signed certificates, the configuration, and the triggers for certificate rotation.
+
+# KubeVirt Certificate Rotation Strategy
+KubeVirt applies self-signed certificate mechamism, where both CA and certifcates are rotated on a defined recurring interval. You can check the corresponding setting  `certificateRotateStrategy` by using cmd
+```sh
+kubectl get kubevirt -n harvester-system -o yaml
+```
+By default, the `certificateRotateStrategy` is empty, meaning KubeVirt will use its default rotation settings, and no manual configuration is needed.
+```yaml
+certificateRotateStrategy: {}
+```
+A complete configuration look like this:
+```yaml
+certificateRotateStrategy:
+  selfSigned:
+    ca:
+      duration: 24h
+      renewBefore: 4h
+    server:
+      duration: 168h
+      renewBefore: 28h
+```
+Explanation of Fields:
+- `.ca.duration`: refers to the validity period the CA certificate. The default value is 7 days.
+- `.ca.renewBefore`: refers to how much time before the CA certificate expiration a new certificate should be issued. The default value is 7 days * 0.2.
+- `.server.duration`: refers to the validity period of server certificates (e.g. virt-api, virt-handler, virt-operator, etc.). The default value is 1 day.
+- `.server.renewBefore`: refers to how much time before the server certificates expire a new certificate should be issued. The default value is 1 day * 0.2.
+
+## Triggers for Certificate Rotation
+The following conditions can trigger certificate rotation in KubeVirt, note that this list highlights key triggers and is not exhaustive.
```

Comment:
when kubevirt rotates the certs, are there a short time-window ? that will cause new VMs are failed to be created?

Reply:
Good point. From what I can see, there is an overlap period where the old and new certs are merged and served, before the old one expired: https://github.com/kubevirt/kubevirt/blob/ec514a29d5552cf5442f262a140681c099e124c8/pkg/virt-operator/resource/apply/core.go#L533-L551

Reply:
Thank you both !
> when kubevirt rotates the certs, are there a short time-window ? that will cause new VMs are failed to be created?

I tried to test this by setting KubeVirt CR `certificateRotateStrategy` to
```yaml
certificateRotateStrategy:
  selfSigned:
    ca:
      duration: 60s
      renewBefore: 10s
    server:
      duration: 30s
      renewBefore: 5s
configuration:
```

Then I tried to pause/unpause vm multiple times, and it turns out during the cert rotation, there might have a chance that KubeVirt may be unable to handle requests.
![image](https://github.com/user-attachments/assets/bce171cf-56dc-4737-8c32-ed784c36ad2d)

Reply:
Perhaps we need to warn users about [the comment above](https://github.com/harvester/harvesterhci.io/pull/73#discussion_r1872811080). I’d appreciate your thoughts on this. Thanks! c.c. @bk201

Reply:
When the side effect is trully existing, we may consider:
(1) Add this formally in the document;
(2) Evaluate a proper default value on Harvester, is current KubeVirt's default value fitting for Harvester? Expose it to Harvester setting so to let user have an option to adjust.
(3) Have a place on the UI to show how long it will expire, user can monitor it, trigger the rotation actively e.g. on a cluster maintainence period.

Reply:
@w13915984028 The PR should be used to clarify the current kubevirt cert rotate behavior, for any improvement it's out of scope and please file an issue to track.